DMCompiler/Optimizer/AnnotatedBytecode.cs

```diff
@@ -0,0 +1,549 @@
+            case OpcodeArgType.StackDelta:
+                return arg is AnnotatedBytecodeStackDelta;
+            case OpcodeArgType.Resource:
+                return arg is AnnotatedBytecodeResource;
+            case OpcodeArgType.TypeId:
+                return arg is AnnotatedBytecodeTypeID;
+            case OpcodeArgType.ProcId:
+                return arg is AnnotatedBytecodeProcID;
+            case OpcodeArgType.FilterId:
+                return arg is AnnotatedBytecodeFilter;
+            case OpcodeArgType.ListSize:
+                return arg is AnnotatedBytecodeListSize;
+            case OpcodeArgType.Int:
+                return arg is AnnotatedBytecodeInteger;
+            case OpcodeArgType.Label:
+                return arg is AnnotatedBytecodeLabel;
+            case OpcodeArgType.Float:
+                return arg is AnnotatedBytecodeFloat;
+            case OpcodeArgType.String:
+                return arg is AnnotatedBytecodeString;
+            case OpcodeArgType.Reference:
+                return arg is AnnotatedBytecodeReference;
+            case OpcodeArgType.FormatCount:
+                return arg is AnnotatedBytecodeFormatCount;
+            case OpcodeArgType.PickCount:
+                return arg is AnnotatedBytecodePickCount;
+            case OpcodeArgType.ConcatCount:
+                return arg is AnnotatedBytecodeConcatCount;
+            default:
+                throw new ArgumentOutOfRangeException(nameof(requiredArg), requiredArg, null);
+        }
+    }
+
+    public void AddArg(IAnnotatedBytecode arg) {
+        _args.Add(arg);
+    }
+
+    public List<IAnnotatedBytecode> GetArgs() {
+        return _args;
+    }
+
+    private Location? _location;
+    public void SetLocation(IAnnotatedBytecode loc) {
+        if (_location != null) return;
+        _location = loc.GetLocation();
+    }
+    public void SetLocation(Location loc) {
+        if (_location != null) return;
+        _location = loc;
+    }
+    public Location GetLocation() {
+        return _location ?? Location;
+    }
+}
+
+internal sealed class AnnotatedBytecodeVariable : IAnnotatedBytecode {
+    public int Exit;
+    public bool Exitingscope;
+    public Location Location;
+    public string Name;
+
+    public AnnotatedBytecodeVariable(string name, Location location) {
+        Name = name;
+        Location = location;
+        Exitingscope = false;
+    }
+
+    public AnnotatedBytecodeVariable(int popoff, Location location) {
+        Exitingscope = true;
+        Exit = popoff;
+        Location = location;
+    }
+
+    public void AddArg(IAnnotatedBytecode arg) {
+        DMCompiler.ForcedError(Location, "Cannot add args to a variable");
+    }
+    public void SetLocation(IAnnotatedBytecode loc) {
+        Location = loc.GetLocation();
+    }
+    public void SetLocation(Location loc) {
+        Location = loc;
+    }
+    public Location GetLocation() {
+        return Location;
+    }
```

Comment:
These methods don't have spacing between them.

Same with every class below this one.